### PR TITLE
Refactored DisplayMemberPath

### LINKED ISSUE #900: [All] DisplayMemberPath not working on custom control template for ComboBox
## I'm submitting a...

- Bug report (I searched for similar issues and did not find one)

## Current behavior

If using a combobox with a style defined in App.xaml the DisplayMemberPath property is not used.  Works if no style is set.

## Expected behavior

DisplayMemberPath should work in Android as on UWP

## Minimal reproduction of the problem with instructions

[App9.zip](https://github.com/nventive/Uno/files/3212928/App9.zip)
Build attached project and test on Android to see issue.  Note that removing the style from the combox resolves the problem.

## Environment

```
Nuget Package:
Uno.UI
Package Version(s):
1.45.0-dev.1524
Affected platform(s):
- [ ] iOS
- [x] Android
- [ ] WebAssembly
- [ ] Windows
- [ ] Build tasks

Visual Studio
- [ ] 2017 (version: )
- [x] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins
- [ ] Resharper (version: )

### LINKED ISSUE #3195: DisplayMemberPath not working on custom control template for ListView
## Current behavior

If using a ListView with a style defined elsewhere the DisplayMemberPath property is not used. Works if no style is set.

## Expected behavior

DisplayMemberPath should work in UWP and WASM while using inside a user control 

## How to reproduce it (as minimally and precisely as possible)
Build attached project and test on WASM to see issue. Note that removing the style from the ListView resolves the problem.

https://filebin.net/hw3sjgiqkwjgj1ig
## Environment

Nuget Package:

Package Version(s):

Affected platform(s):

- [ ] iOS
- [ ] Android
- [ x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [x ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Visual Studio:

- [ ] 2017 (version: )
- [x ] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

### LINKED ISSUE #3545: [WASM] Setting Template for ComboBoxItem breaks use of DisplayMemberPath
## Current behavior

In WASM, if you use a ComboBoxItem style that sets Template then DisplayMemberPath does not work correctly.

## Expected behavior

ComboBox should still display content based on DisplayMemberPath, as in UWP

## How to reproduce it (as minimally and precisely as possible)

Launch the attached project in UWP and observe that the ComboBox, when clicked, displays the values assigned to String1.  Launch the project in WASM and observe that the content is not displayed, just the class name.

[ComboBoxSample.zip](https://github.com/unoplatform/uno/files/4925283/ComboBoxSample.zip)

## Environment

Nuget Package:
Uno.UI
Package Version(s):
3.0.0-dev.783
Affected platform(s):

- [ ] iOS
- [ ] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Visual Studio:

- [ ] 2017 (version: )
- [ ] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

### LINKED ISSUE #3995: ComboBox SelectedValuePath not binding
## Current behavior

The ComboBox binding to Items through the properties ```SelectedValuePath``` ```DisplayMemberPath```

## Expected behavior
The Items should be displayed in the combo box, the selected item showing the text  bound to the ```DisplayMemberValuePath``` and should select the item in the ```SelectedValuePath``` property

## Repro

```xaml
     <ComboBox  ItemsSource="{x:Bind SomeOptions}"
                           SelectedValue="{x:Bind SomeOption, Mode=OneWay}"
                           SelectedValuePath="Option"
                           DisplayMemberPath="Text" />
```

Nuget Package:
- [x] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
Affected platform(s):

- [x] iOS
- [x] Android
- [x] WebAssembly

### PR DESCRIPTION
**Pull request by carldebilly**
* Fixes #900
* Fixes #3195
* Fixes #3545 
* Fixes #3995

# Bugfix
The property `DisplayMemberPath` were not properly implemented in Uno.

## Previous behavior
An internal `DisplayMemberTemplateSelector` was add as template for items when the `DisplayMemberPath` were used.

## New behavior
The `DisplayMemberPath` is now resolved and applied before setting the `DataContext` to an element in an `<ItemsControl>`

## PR Checklist

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [X] Contains **NO** breaking changes
- [X] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [X] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI/Mock/ViewExtensions.netstdref.cs [hunk 1]
```diff
@@ -18,9 +18,9 @@ public static partial class ViewExtensions
 
 		internal static FrameworkElement GetTopLevelParent(this UIElement view) => throw new NotImplementedException();
 
-		internal static T FindFirstChild<T>(this FrameworkElement root) where T : FrameworkElement
+		internal static T FindFirstChild<T>(this FrameworkElement root, bool includeCurrent) where T : FrameworkElement
 		{
-			return root.GetDescendants().OfType<T>().FirstOrDefault();
+			throw new NotImplementedException();
```
Comment (dr1rrb) [inline src/Uno.UI/Mock/ViewExtensions.netstdref.cs:23]: Why did you remove the implementation?

Comment (carldebilly) [reply src/Uno.UI/Mock/ViewExtensions.netstdref.cs:23]: Because it's netstdref and anywhere it's calling `GetDescendants()` that will throw this (right after in the code).

File: src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs [hunk 2]
```diff
@@ -1056,7 +1060,19 @@ protected virtual void PrepareContainerForItemOverride(DependencyObject element,
 					// Set the datacontext first, then the binding.
 					// This avoids the inner content to go through a partial content being
 					// the result of the fallback value of the binding set below.
-					containerAsContentControl.DataContext = item;
+
+					var displayMemberPath = DisplayMemberPath;
+					if (string.IsNullOrEmpty(displayMemberPath))
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/Controls/ItemsControl/ItemsControl.cs:1065]: You should make this available to the `containerAsContentPresenter` code path as well - this is hit when using a bare `<ItemsControl \>`.

File: src/Uno.UI/Mock/ViewExtensions.netstdref.cs (hunk 1 above)
Comment (dr1rrb) [reply src/Uno.UI/Mock/ViewExtensions.netstdref.cs:23]: Ok for the "netstdref", otherwise I would say that the fact `GetDescendants()` is not implemented is not a valid reason. If we gonna implement it, there was about 101% of chances that we forgot to re-implement this method.

File: src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ItemsControl.cs [hunk 3]
```diff
@@ -66,6 +68,55 @@ public async Task When_ContainerSet_Then_ContentShouldBeSet()
...
+			{
+				new ItemForDisplayMemberPath {DisplayName = "item 0"},
+				new ItemForDisplayMemberPath {DisplayName = "item 1"},
+				new ItemForDisplayMemberPath {DisplayName = "item 2"}
+			};
+
+			SUT.ItemsSource = source;
+
+
+			async Task Assert(int index, string s)
```
[the remark this answers is not in the document]
Comment (carldebilly) [reply src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_ItemsControl.cs:103]: Oooh... that's a real bug.  I'll fix it.

**Review by davidjohnoliver — APPROVED** [review]
